numpyro/infer/hmc_gibbs.py

```diff
@@ -0,0 +1,109 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import namedtuple
+import copy
+from functools import partial
+
+from jax import device_put, random, value_and_grad
+
+from numpyro.handlers import condition, seed, trace, substitute
+from numpyro.infer.mcmc import MCMCKernel
+from numpyro.infer.hmc import HMC
+
+
+HMCGibbsState = namedtuple("HMCGibbsState", "z, hmc_state, rng_key")
+"""
+ - **z** - a dict of the current latent values (both HMC and Gibbs sites)
+ - **hmc_state** - current hmc_state
+ - **rng_key** - random key for the current step
+"""
+
+
+def _wrap_model(model):
+    def fn(*args, **kwargs):
+        gibbs_values = kwargs.pop("_gibbs_sites", {})
+        with condition(data=gibbs_values), substitute(data=gibbs_values):
+            model(*args, **kwargs)
+    return fn
+
+
+class HMCGibbs(MCMCKernel):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    HMC-within-Gibbs. This inference algorithm allows the user to combine
+    general purpose gradient-based inference (HMC or NUTS) with custom
+    Gibbs samplers.
+
+    :param inner_kernel: One of :class:`~numpyro.infer.HMC` or :class:`~numpyro.infer.NUTS`.
+    :param gibbs_fn: A Python callable that returns a dictionary of Gibbs samples conditioned
+        on the HMC sites. Must include an argument `rng_key` that should be used for all sampling.
+        Must also include arguments for all HMC and Gibbs sites.
+    :param gibbs_sites: a list of site names for the latent variables that are covered by the Gibbs sampler.
+    """
+
+    sample_field = "z"
+
+    def __init__(self, inner_kernel, gibbs_fn, gibbs_sites):
+        if not isinstance(inner_kernel, HMC):
+            raise ValueError("inner_kernel must be a HMC or NUTS sampler.")
+        if not callable(gibbs_fn):
+            raise ValueError("gibbs_fn must be a callable")
```

Comment:
We need to assert `inner_kernel.model is not None` here. It is trickier to support `potential_fn` input IMO.